adaptive/learner/new_learnerND.py

```diff
@@ -0,0 +1,513 @@
+        "Remove and return the item and priority at the front of the queue."
+        (key, item) = self._queue.popitem()
+        i = self._items.index((item, key))
+        del self._items[i]
+        priority, _ = key
+        return item, priority
+
+    def insert(self, item, priority):
+        "Insert 'item' into the queue with the given priority."
+        key = (priority, self._n)
+        self._items.add((item, key))
+        self._queue[key] = item
+        self._n += 1
+
+    def remove(self, item):
+        "Remove the 'item' from the queue."
+        i = self._items.bisect_left((item, ()))
+        should_be, key = self._items[i]
+        if item != should_be:
+            raise KeyError("item is not in queue")
+
+        del self._queue[key]
+        del self._items[i]
+
+    def update(self, item, new_priority):
+        """Update 'item' in the queue with the given priority.
+
+        Raises
+        ------
+        KeyError : if 'item' is not in the queue.
+        """
+        i = self._items.bisect_left((item, ()))
+        should_be, key = self._items[i]
+        if item != should_be:
+            raise KeyError("item is not in queue")
+
+        _, n = key
+        new_key = (new_priority, n)
+
+        del self._queue[key]
+        self._queue[new_key] = item
+        del self._items[i]
+        self._items.add((item, new_key))
+
+
+class LossFunction:
+    @property
+    def n_neighbors(self):
+        "The maximum degree of neighboring subdomains required."
+
+    def __call__(self, domain, subdomain, data):
+        """Return the loss for 'subdomain' given 'data'
+
+        Neighboring subdomains can be obtained with
+        'domain.neighbors(subdomain, self.n_neighbors)'.
+        """
+
+
+class DistanceLoss(LossFunction):
+    @property
+    def n_neighbors(self):
+        return 0
+
+    def __call__(self, domain, subdomain, codomain_bounds, data):
+        # XXX: this is specialised to 1D
+        a, b = subdomain
+        ya, yb = data[(a,)], data[(b,)]
+        return sqrt((b - a) ** 2 + (yb - ya) ** 2)
+
+
+def _scaled_loss(loss, domain, subdomain, codomain_bounds, data):
+    subvolumes = domain.subvolumes(subdomain)
+    max_relative_subvolume = max(subvolumes) / sum(subvolumes)
+    L_0 = loss(domain, subdomain, codomain_bounds, data)
+    return max_relative_subvolume * L_0
+
+
+class LearnerND(BaseLearner):
+    def __init__(self, f, bounds, loss=None):
+
+        if len(bounds) == 1:
+            (a, b), = bounds
+            self.domain = Interval(a, b)
+            self.loss = loss or DistanceLoss()
+        else:
+            raise ValueError("Can only handle 1D functions for now")
+
+        self.queue = Queue()
+        self.data = dict()
+        self.function = f
+
+        # Evaluate boundary points right away to avoid handling edge
+        # cases in the ask and tell logic
+        bound_points = sorted(map(tuple, itertools.product(*bounds)))
+        for x in bound_points:
+            y = f(x)
```

Comment:
Does this really need to happen in `__init__`?

Reply:
It simplifies everything else if we do. We can lift this restriction at the end, but I'd rather not complicate the implementation of `ask` for the moment.